### PR TITLE
Implementation of additional external metrics for classification/clustering

### PR DESCRIPTION
**Pull request by hoanganhngo610**
This is the implementation of various external metrics that can solve the problem of permutation of clustering/classification labels, including:

- Homogeneity, Completeness and VBeta
- Mutual Information, including Mutual Information, Adjusted Mutual Information, Expected Mutual Info and Normalized Mutual Information
- Matthews Correlation Coefficient
- Prevalence Threshold
- Purity
- Q0 and Q2 as a normalization of Q0
- Purity
- Rand Index and Adjusted Rand Index
- Sorensen-Dice index based on the currently available Jaccard Index

### CODE REVIEW

File: river/metrics/rand.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        except ZeroDivisionError:
+            return 0.0
+
+
+class AdjustedRand(base.Metric):
+    """Adjusted Rand Index.
+
+    The Adjusted Rand Index is the corrected-for-chance version of the Rand Index [^1] [^2].
+    Such a correction for chance establishes a baseline by using the expected similarity
+    of all pair-wise comparisions between cluserings specified by a random model.
```
Comment (hoanganhngo610) [inline river/metrics/rand.py:121]: `clusteings` modified to `clusterings`. Will be updated when the review is complete.

**Comment by smastelini** [comment]
Hi @hoanganhngo610, could you please update the PR description and title to reflect everything that you have implemented so far?

**Comment by hoanganhngo610** [comment]
@smastelini Thank you so much for your reminder! Currently, I'm still deciding on which metrics to be implemented, so the description will be updated on the go while new metrics are added! The description will be updated tmr the latest.

File: river/metrics/mutual_info.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+    where
+
+    * $(a_i + b_j - N)^+$ denotes $\max(1, a_i + b_j - N)$,
+
+    * $a_i$ is the sum of row i-th of the contingency table, and
+
+    * $b_j$ is the sum of column j-th of the contingency table.
+
+    The Expected Mutual Information score will be used to calculate the Adjusted Mutual
```
Comment (jacobmontiel) [inline river/metrics/mutual_info.py:269]: ~~This note should be in the AMI class.~~

Ignore this, I just saw the note in AMI.

Comment (jacobmontiel) [reply river/metrics/mutual_info.py:269]: ~~Also, consider the following rewording:

> The Adjusted Mutual Information score (AMI) relies on the Expected Mutual Information 
   (EMI) score. Due to the complexity of EMI, the AMI is one order of magnitude slower than 
    other metrics.~~

Comment (jacobmontiel) [reply river/metrics/mutual_info.py:269]: Replace with a note here mentioning that the metric is very expensive to calculate.

File: river/metrics/mutual_info.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+    However, due to the complexity of the Expected Mutual Info Score, the computation of
+    this metric is an order of magnitude slower than most other metrics, in general.
+
+    Parameters
+    ----------
+    cm
+        This parameter allows sharing the same confusion
+        matrix between multiple metrics. Sharing a confusion matrix reduces the amount of storage
+        and computation time.
+
```
Comment (jacobmontiel) [inline river/metrics/mutual_info.py:394]: Missing parameter `average_method`

File: river/metrics/pair_confusion.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    * $C[1][1]$ (**True Positives**): number of pairs of points that are in the same cluster
+    in both true and predicted labels.
+
+    We can also show that the four counts have the following property
+
+    $$
+    TP + FP + FN + TV = \frac{n(n-1)}{2}
+    $$
+
+    Parameters
```
Comment (jacobmontiel) [inline river/metrics/pair_confusion.py:34]: Duplicated text

File: river/metrics/pair_confusion.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        for i in self.cm.classes:
+            for j in self.cm.classes:
+                sum_squares += self.cm[i][j] * self.cm[i][j]
+
+        true_positives = sum_squares - self.cm.n_samples
+
+        false_positives = 0
+        for i in self.cm.classes:
+            for j in self.cm.classes:
+                try:
```
Comment (jacobmontiel) [inline river/metrics/pair_confusion.py:87]: This calculation can go inside the loop above. No need to loop again.

File: river/metrics/pair_confusion.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                    false_positives += self.cm[i][j] * self.cm.sum_col[j]
+                except KeyError:
+                    continue
+        false_positives = false_positives - sum_squares
+
+        false_negatives = 0
+        for i in self.cm.classes:
+            for j in self.cm.classes:
+                try:
```
Comment (jacobmontiel) [inline river/metrics/pair_confusion.py:96]: Same comment, no need to loop again as this can go inside the first loop.

File: river/metrics/prevalence_threshold.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    The relationship between a positive predicted value and its target prevalence
+    is propotional - though not linear in all but a special case. In consequence,
+    there is a point of local extrema and maximum curvature defined only as a function
+    of the sensitivity and specificity beyond which the rate of change of a test's positive
+    predictive value drops at a differential pace relative to the disease prevalence.
+    Using differential equations, this point was first defined by Balayla et al. [^1] and
+    is termed the **prevalence threshold** (\phi_e).
+
+    The equation for the prevalence threshold [^2] is given by the following formula
```
Comment (jacobmontiel) [inline river/metrics/prevalence_threshold.py:19]: Sounds redundant.

File: river/metrics/q0.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    The motivation for this approach is an appeal to parsimony: Given identical
+    conditional entropies, H(C|K), the clustering solution with the fewest clusters
+    should be preferred.
+
+    The Q0 measure can be calculated using the following formula [^1]
+
+    $$
+    Q_0(C, K) = H(C|K) + \frac{1}{n} \sum_{k=1}^{|K|} \log \binom{h(c) + |C| - 1}{|C| - 1}.
+    $$
+
```
Comment (jacobmontiel) [inline river/metrics/q0.py:29]: Should we consider a note regarding the complexity of this metric?

File: river/metrics/rand.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+        true_positives = pair_confusion_matrix[1][1]
+        true_negatives = pair_confusion_matrix[0][0]
+
+        total_pairs = self.cm.n_samples * (self.cm.n_samples - 1)
+
+        try:
+            return (true_positives + true_negatives) / total_pairs
+        except ZeroDivisionError:
+            return 1.0
```
Comment (jacobmontiel) [inline river/metrics/rand.py:91]: Why 1.0? Should this metric define:

```python
@property
    def bigger_is_better(self):
        return False
```

File: river/metrics/rand.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+                2.0
+                * (true_positives * true_negatives - false_negatives * false_positives)
+                / (
+                    (true_positives + false_negatives)
+                    * (false_negatives + true_negatives)
+                    + (true_positives + false_positives)
+                    * (false_positives + true_negatives)
+                )
+            )
+        except ZeroDivisionError:
```
Comment (jacobmontiel) [inline river/metrics/rand.py:174]: Same comment as for Rand Index.

File: river/metrics/test_.py [hunk 11]
```diff
@@ -144,6 +144,22 @@ def partial(f, **kwargs):
...
+    (metrics.Homogeneity(), sk_metrics.homogeneity_score),
+    (metrics.Completeness(), sk_metrics.completeness_score),
+    (metrics.VBeta(beta=0.5), partial(sk_metrics.v_measure_score, beta=0.5)),
+    (metrics.MatthewsCorrCoef(), sk_metrics.matthews_corrcoef),
+    (metrics.FowlkesMallows(), sk_metrics.fowlkes_mallows_score),
+    (metrics.Rand(), sk_metrics.rand_score),
+    (metrics.AdjustedRand(), sk_metrics.adjusted_rand_score),
+    (metrics.MutualInfo(), sk_metrics.mutual_info_score),
+    (
+        metrics.NormalizedMutualInfo(average_method="geometric"),
```
Comment (jacobmontiel) [inline river/metrics/test_.py:156]: What about testing the other average_methods?

Comment (jacobmontiel) [reply river/metrics/test_.py:156]: Same for AMI

File: river/metrics/vbeta.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+                continue
+
+        try:
+            return 1.0 - raw_conditional_entropy / max_reduction_entropy
+        except ZeroDivisionError:
+            return 1.0
+
+
+class VBeta(metrics.MultiClassMetric):
+    r"""V-Measure.
```
Comment (jacobmontiel) [inline river/metrics/vbeta.py:210]: Does not match the class name

**Review by jacobmontiel — CHANGES_REQUESTED** [review]
Awesome work @hoanganhngo610 

I left a couple of comments regarding minor changes.

File: river/metrics/base.py [hunk 13]
```diff
@@ -334,6 +348,10 @@ def works_with(self, model) -> bool:
     def bigger_is_better(self):
         raise NotImplementedError
 
+    @property
+    def works_with_weights(self):
+        return NotImplementedError
```
[the remark this answers is not in the document]
Comment (hoanganhngo610) [reply river/metrics/base.py:353]: Resolved! Implemented (similar to `works_with`)

```python
return all(m.works_with_weights for m in self)
```

File: river/metrics/fowlkes_mallows.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+
+class FowlkesMallows(base.MultiClassMetric):
+    r"""Fowlkes-Mallows Index.
+
+    The Fowlkes-Mallows Index [^1] [^2] is an external evaluation method that is
+    used to determine the similarity between two clusterings, and also a mmetric
+    to measure confusion matrices. The measure of similarity could be either between
+    two hierarchical clusterings or a clustering and a benchmark classification. A
+    higher value for teh Fowlkes-Mallows index indicates a greater similarity between
```
[the remark this answers is not in the document]
Comment (hoanganhngo610) [reply river/metrics/fowlkes_mallows.py:17]: Resolved!

File: river/metrics/fowlkes_mallows.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    the clusters and the benchmark classifications.
+
+    The Fowlkes-Mallows Index, when results of two cluster algorithms are used to
```
[the remark this answers is not in the document]
Comment (hoanganhngo610) [reply river/metrics/fowlkes_mallows.py:20]: Resolved!

File: river/metrics/matthews_corrcoef.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    Machine Learning as a measure of the quality of classifications, introduced
+    by Brian W. Matthews in 1975. The MCC is defined identically to Pearson's phi
+    coefficient, introduced by Karl Pearson, also known as the Yule phi coefficient
+    from its introduction by Udny Yule in 1912.
+
+    The coefficient takes into account true and false positives and negatives and is
+    generally regarded as a balanced measure which can be used even if the classes
+    are of very different sizes. It returns a value between -1 and 1, with 1 being
+    a perfect prediction, 0 no better than random prediction and -1 means a total
+    disagrrement between prediction and observation.
```
[the remark this answers is not in the document]
Comment (hoanganhngo610) [reply river/metrics/matthews_corrcoef.py:21]: Resolved!

File: river/metrics/matthews_corrcoef.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+          Plos ONE, 7(8), e41882. doi: 10.1371/journal.pone.0041882
+
+    """
+
+    def __init__(self, cm=None):
+        super().__init__(cm)
+
+    def get(self):
+
+        n_correct = sum(self.cm[i][i] for i in self.cm.classes)
```
[the remark this answers is not in the document]
Comment (hoanganhngo610) [reply river/metrics/matthews_corrcoef.py:84]: Resolved!

File: river/metrics/mutual_info.py [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+    "MutualInfo",
+    "NormalizedMutualInfo",
+]
+
+
+class MutualInfo(metrics.MultiClassMetric):
+    r"""Mutual Information between two clusterings.
+
+    The Mutual Information [^1] is a measure of the similarity between two labels of
+    the same data. Where $|U_i|$ is the number of samples in cluster $U_i$ and $V_j$
```
[the remark this answers is not in the document]
Comment (hoanganhngo610) [reply river/metrics/mutual_info.py:20]: Resolved!

File: river/metrics/mutual_info.py [hunk 19]
```diff
@@ -0,0 +1,498 @@
...
+    is the number of the samples in cluster $V_j$, the Mutual Information between
+    clusterings $U$ and $V$ can be calculated as:
+
+    $$
+    MI(U,V) = \sum_{i=1}^{|U|} \sum_{v=1}^{|V|} \frac{|U_i \cup V_j|}{N} \log \frac{N |U_i \cup V_j|}{|U_i| |V_j|}
+    $$
+
+    This metric is independent of the absolute values of the labels: a permutation
+    of the class or cluster label values won't change the score in any way possible.
```
[the remark this answers is not in the document]
Comment (hoanganhngo610) [reply river/metrics/mutual_info.py:29]: Resolved!

File: river/metrics/mutual_info.py [hunk 20]
```diff
@@ -0,0 +1,498 @@
...
+    where $H(U)$ and $H(V)$ are the marginal entropies, $H(U | V)$ and $H(V | U)$ are the
+    conditional entropies.
+
+    Parameters
+    ----------
+    cm
+        This parameter allows sharing the same confusion
+        matrix between multiple metrics. Sharing a confusion matrix reduces the amount of storage
+        and computation time.
+
```
Comment (hoanganhngo610) [inline river/metrics/mutual_info.py:148]: Added `average_method`as a parameter.

File: river/metrics/mutual_info.py (hunk 3 above)
Comment (hoanganhngo610) [reply river/metrics/mutual_info.py:394]: Resolved!

File: river/metrics/pair_confusion.py (hunk 4 above)
Comment (hoanganhngo610) [reply river/metrics/pair_confusion.py:34]: Resolved!

File: river/metrics/pair_confusion.py (hunk 5 above)
Comment (hoanganhngo610) [reply river/metrics/pair_confusion.py:87]: Resolved!

File: river/metrics/pair_confusion.py (hunk 6 above)
Comment (hoanganhngo610) [reply river/metrics/pair_confusion.py:96]: Resolved!

File: river/metrics/vbeta.py (hunk 12 above)
Comment (hoanganhngo610) [reply river/metrics/vbeta.py:210]: Resolved!

File: river/metrics/test_.py (hunk 11 above)
Comment (hoanganhngo610) [reply river/metrics/test_.py:156]: I initially thought that repeating the same test for different average methods would be a little bit redundant, but if needed, I would add the test now!

Comment (hoanganhngo610) [reply river/metrics/test_.py:156]: Resolved!

File: river/metrics/rand.py (hunk 9 above)
Comment (hoanganhngo610) [reply river/metrics/rand.py:91]: When the total number of pairs is zero, the rand index is automatically set to `1.0`. This is done so that the results are 100% in line with the implementation of the rand index in scikit-learn. 
And yes, the `bigger_is_better` should definitely be implemented!

File: river/metrics/rand.py (hunk 10 above)
Comment (hoanganhngo610) [reply river/metrics/rand.py:174]: Same reasoning as above!

File: river/metrics/mutual_info.py [hunk 21]
```diff
@@ -0,0 +1,498 @@
...
+                except (ValueError, ZeroDivisionError):
+                    continue
+
+        return mutual_info_score
+
+
+class NormalizedMutualInfo(metrics.MultiClassMetric):
+    r"""Mutual Information between two clusterings.
+
+    The Mutual Information [^1] is a measure of the similarity between two labels of
```
[the remark this answers is not in the document]
Comment (hoanganhngo610) [reply river/metrics/mutual_info.py:116]: Resolved!

File: river/metrics/mutual_info.py (hunk 2 above)
Comment (hoanganhngo610) [reply river/metrics/mutual_info.py:269]: Resolved!

File: river/metrics/prevalence_threshold.py (hunk 7 above)
Comment (hoanganhngo610) [reply river/metrics/prevalence_threshold.py:19]: I think the paragraph is necessary to explain the existence of the metric. Otherwise, it would be a bit difficult to understand how it will be applied.

File: river/metrics/q0.py (hunk 8 above)
Comment (hoanganhngo610) [reply river/metrics/q0.py:29]: I have put a note here, and in the normalized version of the metric (Q2), saying that this is one order of magnitude slower than most other implemented metrics.

**Review by hoanganhngo610 — COMMENTED** [review]
@jacobmontiel Thank you so much for your comments! I have updated the code according to your recommendations!

**Review by jacobmontiel — APPROVED** [review]
Changes are correct, thanks @hoanganhngo610 !

The PR is ready to merge.

File: river/metrics/rand.py (hunk 9 above)
Comment (jacobmontiel) [reply river/metrics/rand.py:91]: Understood, thanks for the clarification.

**Comment by jacobmontiel** [comment]
The mutual information test fails due to differences in the implementation with `sklearn`. It has been decided to align with `sklearn` for consistency.

**Review by jacobmontiel — APPROVED** [review]
Thanks for the hard work @hoanganhngo610 

The PR is ready.
